### ISSUE TITLE
False-negative "Match must be exhaustive" when matching pair of references

### ISSUE THREAD

**Issue by dima74** [intellij-rust/intellij-rust#6131]
## Environment

* **IntelliJ Rust plugin version:** 0.3.132.3356-202-nightly
* **Rust toolchain version:** 1.46.0 (04488afe3 2020-08-24) x86_64-unknown-linux-gnu
* **IDE name and version:** IntelliJ IDEA 2020.2.1 Community Edition (IC-202.6948.69)
* **Operating system:** Linux 5.8.8-arch1-1

## Problem description
There is no inspection "Match must be exhaustive" on this code when using version 0.3.132.3356-202-nightly:
![image](https://user-images.githubusercontent.com/6505554/93530517-5bb77c80-f946-11ea-962f-cf4eb0b00fbf.png)

There is inspection when using version 0.3.130.3337-202:
![image](https://user-images.githubusercontent.com/6505554/93530753-c4065e00-f946-11ea-975f-91c9218b50b5.png)

## Steps to reproduce
```rust
enum E { A, B }
use E::*;
fn foo(a: &E, b: &E) {
    match (a, b) {
        (A, A) => {}
        (_, B) => {}
    }
}
```

Probably this regression was introduced in #5720

Related issue: #5901 (it was fixed by #5720)

**Comment by ortem** [comment]
You're right, this was introduced in #5720.

In order to fix some false positives related to match ergonomics or mismatched types inside match expresssions (e.g. #5901), `Non-exhaustive match` inspection now stops on any match expression with ambigous type inside (this is handled by `CheckMatchUtilsKt#getFirstColumnType`). The ambiguity in this case is caused by match ergonomics in type inference:

```Rust
fn foo(a: &E, b: &E) {
    match (a, b) {
        (A/*type=E*/, A/*type=E*/) => {}
        (_/*type=&E*/, B/*type=E*/) => {}
    }
}
```
there are two different types `E` and `&E` inside so the inspection stops and doesn't highlight the error. The inspection should distinguish such cases from the cases with mismatched types inside and handle them properly.
